### ISSUE TITLE
Parsing less than complete ISO representations

### ISSUE THREAD

**Issue by mj1856** [moment/moment#3918]
`moment('2017-01')` is undocumented in the [supported ISO strings](http://momentjs.com/docs/#/parsing/string/), but is correctly parsed as `2017-01-01` local time without giving a deprecation warning.

`moment('201701')` is the ISO 8601 basic form of the same representation, but this gives a deprecation warning, falls back to JS Date, and results in the entire number being interpreted as the year.

`moment('2017')` is also allowed under ISO8601, but we don't support it.

We should either fully support minimal representations less than a full date, or not support them at all, and we should document this either way.

**Comment by butterflyhug** [comment]
Are you sure that '201701' is a valid ISO basic representation?  We have a comment in the code that asserts YYYYMM (aka '201701') is not part of the standard:  https://github.com/moment/moment/blob/7589bb8967aca2bda0192595e6e53a12c750aa69/src/lib/create/from-string.js#L22

Assuming past-us is correct here, I think it would minimize confusion if we support and document '2017-01' and '2017', and explicitly document that '201701' was omitted from the standard.

**Comment by alburthoffman** [comment]
ISO 8601 does suport simply form '201701' and '2017'
we should fix this one, and add test cases for this.

**Comment by butterflyhug** [comment]
I repeat:  are you *sure* about that?

I don't have a copy of the standard handy, but Wikipedia also asserts that year+month requires a hyphen alongside a plausible rationale:

> By disallowing dates of the form YYYYMM, the standard avoids confusion with the truncated representation YYMMDD (still often used).

https://en.m.wikipedia.org/wiki/ISO_8601#Calendar_dates

**Comment by mj1856** [comment]
You know what, you're right.  I have a copy, and for year+month, the *basic* form is `YYYY-MM` and the extended form is listed as "not applicable".

Year-only *is* allowed though.  It is at least positive four digits 0000 - 9999, but can support five or 6 digits and negatives "by agreement"

We only support positive four-digit years with `YYYY-MM-DD`, so I think if we add year-only recognition it should match.

To summarize:

- Document that we do support `YYYY-MM` as a recognized ISO8601 string
- Add support for `YYYY` and document that too.

**Comment by hikikoMOURI** [comment]
Hello! It's my first time contributing. Can I take this? :)

**Comment by medinarkhov** [comment]
Hopefully https://github.com/moment/moment/pull/4470 will fix the issue
